### ISSUE TITLE
Fix FP in S6442 improve corner case and fix rule title

### ISSUE THREAD

**Issue by gabriel-vivas-sonarsource** [SonarSource/SonarJS#3516]
| Label | Description |
| --- | --- |
|Rule | S6442 "React's useState used directly in render function" |
| Reproduce | False positive on some edge cases where a derived state is intended. |

## Explanation

As[ pointed out](https://twitter.com/gigonake/status/1584901827165884418?s=20&t=GDxDIFW7PbZQCUXEM7-I7A) by a user, there are some [corner cases](https://beta.reactjs.org/apis/react/useState#storing-information-from-previous-renders) where you might legitimately want to call the state setter in the render, although inside a condition to avoid the infinite loop.

React discourages this pattern ("This pattern can be hard to understand and is usually best avoided"), but I guess it is a false positive nonetheless.

## Suggestion

I suggest we ignore cases where the state setter is inside a conditional.
For example:

```js
    const [language, setLanguage] = useState("fr-FR");

    if (enjoySharpKnives) {
        setLanguage(navigator.language);
    }
```

Perhaps we can go all the way and require an explicit return statement since this is what React is fundamentally doing internally.

```js
    if (enjoySharpKnives) {
        setLanguage(navigator.language);
        return null;
    }
```

## Corrections

The title of the rule says the opposite of what we want:
```diff
- React's useState hook should only be used in the render function or body of a component
+ React's useState hook should not be used directly in the render function or body of a component
```

While we are there, there are also extra semicolons in the code examples:
```diff
function ShowLanguage() {
    ...
-};
+}
```
